Implement user condition types

Cyperghost opened the pull request:
Ref https://github.com/WoltLab/WCF/pull/6304
Required https://github.com/WoltLab/WCF/issues/6318

Cyperghost committed:
Sort condition types by name

Cyperghost committed:
Add user condition types for email, language, avatar, and signature

Cyperghost committed:
Implement an `AbstractUserBooleanConditionType` and coverPhoto condition type for users

Cyperghost committed:
Add abstract user string condition type and refactor email/username conditions

Cyperghost committed:
Improve sorting logic

Cyperghost committed:
Rename 'id' to 'identifier' in AbstractUserBooleanConditionType

Cyperghost committed:
Fix condition handling in UserRegistrationDaysConditionType

Cyperghost committed:
Add user condition types for banned, email confirmation, and enabled status

Cyperghost committed:
Add user trophy condition types

Cyperghost committed:
Add user integer condition types for activity points, likes received, and trophy points

Cyperghost committed:
Fix SQL query syntax and add TODO for user options conditions

Cyperghost committed:
Refactor user condition types to extend AbstractUserIsNullConditionType and update filter logic

Cyperghost committed:
Refactor user condition types to use anonymous classes for improved flexibility

Cyperghost committed:
Add multifactor authentication condition to user conditions

Cyperghost committed:
Add IMigrateConditionType interface and implement migration methods in user group condition types

Cyperghost committed:
Implement migration support for user condition types

Cyperghost committed:
Fix condition logic for user signature and null checks

Cyperghost committed:
Add migration support for user group assignment conditions

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/AbstractUserIntegerConditionType.class.php:
```diff
@@ -0,0 +1,91 @@
+            '>' => $object->{$this->columnName} < $this->filter['value'],
+            '<' => $object->{$this->columnName} > $this->filter['value'],
+            '>=' => $object->{$this->columnName} <= $this->filter['value'],
+            '<=' => $object->{$this->columnName} >= $this->filter['value'],
```
Are you sure this is correct? If the condition is set to be something like `> 3`, I expected the condition to be `column > 3`, not `column < 3`?

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/UserHasNotTrophyConditionType.class.php:
```diff
@@ -0,0 +1,90 @@
+            "{$objectList->getDatabaseTableAlias()}.userID NOT IN (
+                    SELECT userID
+                    FROM   wcf1_user_trophy
+                    WHERE  trophyID IN = ?
```
```suggestion
                    WHERE  trophyID = ?
```

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/UserHasNotTrophyConditionType.class.php:
```diff
@@ -0,0 +1,90 @@
+        $userTrophies = UserTrophyList::getUserTrophies([$object->userID], false)[$object->userID];
+        $trophyIDs = \array_map(static function ($userTrophy) {
+            return $userTrophy->trophyID;
+        }, $userTrophies);
```
```suggestion
        $trophyIDs = \array_column($userTrophies, 'trophyID');
```

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/user/UserHasTrophyConditionType.class.php:
```diff
@@ -0,0 +1,90 @@
+        $userTrophies = UserTrophyList::getUserTrophies([$object->userID], false)[$object->userID];
+        $trophyIDs = \array_map(static function ($userTrophy) {
+            return $userTrophy->trophyID;
+        }, $userTrophies);
```
```suggestion
        $trophyIDs = \array_column($userTrophies, 'trophyID');
```

Cyperghost committed:
Apply suggestions from code review

Co-authored-by: Alexander Ebert <ebert@woltlab.com>

Cyperghost committed:
Fix comparison operators

Cyperghost committed:
Add missing `.` in the sql query

Cyperghost committed:
Fix comparison logic

Cyperghost committed:
Fix matches logic

Cyperghost committed:
Change Filter Type to string when `SelectFormField` is used

Cyperghost committed:
Escape special characters

Cyperghost committed:
Compare strings case-insensitive

Cyperghost committed:
Fix condition comparison logic

Cyperghost committed:
Use a unique structure for the object types for migration.

Cyperghost committed:
Rename 'needMigration' to 'isLegacy'

Cyperghost committed:
Apply suggestions from code review

Co-authored-by: Alexander Ebert <ebert@woltlab.com>

Cyperghost committed:
Merge remote-tracking branch 'origin/6.3-condition-migrate' into 6.3-condition-migrate

Cyperghost committed:
Refactor `ConditionMigration`

Cyperghost committed:
Refactor migration condition checks for consistency

Cyperghost committed:
Refactor condition data migration logic to improve null handling

Cyperghost committed:
Migrate user condition types to use named parameters for improved readability

Cyperghost committed:
Add legacy notice for automatic user group assignments and implement migration worker

Cyperghost committed:
Rename `UserGroupAssignmentMigrateCondition` to `MigrateLegacyCondition`

Cyperghost committed:
Fix typo in `$field`

Cyperghost committed:
Add error handling for JSON decoding in condition migration

Cyperghost committed:
Fix incorrect usage of unset function in ConditionHandler

Cyperghost committed:
Use `DateTimeImmutable` instead of `\DateTime`

Cyperghost committed:
Expand the PHPDoc for methods in `IMigrateConditionType`

Cyperghost committed:
Apply suggestions from code review

Co-authored-by: Alexander Ebert <ebert@woltlab.com>

Cyperghost committed:
Add a status box if content needs to be recalculated using Rebuild Data.

Cyperghost committed:
Invert check if migrations needed

dtdesign committed:
Avoid inheriting from concrete implementations

dtdesign committed:
Fix the class naming to be consistent

dtdesign committed:
`\strtolower()` is not binary-safe

dtdesign committed:
Fix the handling of condition types

dtdesign committed:
Mark the abstract user conditions as non-abstract

If we're going to instantiate them via anonymous classes, we can just drop the `abstract` because their entire implementation is non-abstract anyway.

dtdesign committed:
Register the conditions sequentually

There is very little gain in offering `addConditions()` plus it requires us to validate an arbitrary array in contrast to relying on `addCondition()` with its type hin.

dtdesign committed:
Fix the direction of the comparison

dtdesign committed:
Simplify the condition names

Avoiding the repetition makes it easier to find conditions.

Cyperghost committed:
Rename the class to the correct name
Add phpstan docs

Cyperghost committed:
Adding language variable
The content of `wcf.condition.add` has been swapped between German and English.

Cyperghost committed:
Merge pull request #6330 from WoltLab/6.3-condition-migrate

Migrate "old" conditions
